ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+
+### Architecture
+
+The workspace team has factored these limitations in while developing features on Konflux downstream by creating a layer between the Konflux User Interfaces and the KubeSaw components that provide namespace as a service (NSaaS) layer. We understand that there are multiple ways to implement a namespace provisioning service depending on the constraints of the system. Our first implementation is KubeSaw, but as we redefined some constraints on Konflux deployments, other backends for a workspace became viable. See the diagram below for the different backends that can be implemented at the moment.
+
+#### Diagram
+
+![](../diagrams/ADR-0040/konflux-workspaces.png)
+
+#### Description
+
+The diagram depicts 3 backend implementations for the Workspace Controller: KubeSaw, Local and GitOps. All backends end up managing namespaces, their initial resources required, and connectivity between these namespaces to authenticated identities.
+
+##### KubeSaw
+
+This is the current implementation downstream running today in our KubeSaw deployments. It leverages the KubeSaw host operator, member operator and registration service components. It is designed to scale with multiple member cluster support. This implementation will only be possible with OpenShift clusters, since there is no vanilla support of K8S at the moment. Implementing support for non-OpenShift Kubernetes deployments is scheduled for the future.
+
+##### Local
+
+This implementation assumes Konflux runs in single cluster mode. It is a straightforward implementation of local namespace provisioning, bootstrapping and user allocation. It will require a workspace config object that describes the default resources to set during the bootstrapping phase.
+
+##### GitOps
+
+This implementation provisions the namespace in a target cluster, but the bootstrapping is delegated to a GitOps provider like ArgoCD. A new application will be spawned on the git repo and ArgoCD will track those resources and synchronize them with the cluster. This is the current approach using upstream at the moment.
+
+### Features by backend
+
+The table below expands with more details on the features that each backend can provide and which features will not be available.
+
+| Feature | KubeSaw | Local | GitOps |
+| :---- | :---- | :---- | :---- |
+| Single-cluster | ✅ | ✅ | ⚠️ (ArgoCD) |
+| Multi-cluster | ✅ | ❌ | ✅ |
+| User Authentication | ⚠️ (IdP) | ⚠️ (IdP) | ⚠️ (IdP) |
+| User Lifecycle | ✅ | ⚠️ (IdP) | ⚠️ (IdP) |
+| Workspace Sharing | ✅ | ⚠️ (K8S) | ⚠️ (K8S) |
+| Self Healing | ✅ | ❌ | ⚠️ (ArgoCD) |
+| Proxying  | ✅ | ❌ | ❌ |
+| Public Workspace | ✅ | ⚠️ (IdP+K8S) | ⚠️ (IdP+K8S) |
+| Workspace Provisioning | ✅ | ✅ | ✅ |
+| Workspace Bootstrapping | ✅ | ✅ | ⚠️ (ArgoCD) |
+| Personal Workspaces | ✅ | ❌ | ❌ |
+| Workspace Idling | ✅ | ❌ | ❌ |
+| Multiple Workspaces | ✅ | ✅ | ✅ |
+| Vanilla K8S | ❌\* | ✅ | ✅ |
+
+\* planned for future
+
+#### Legend
+
+* ✅ implemented in the backend and available  
```

Comment:
The local implementation is already available in https://github.com/konflux-workspaces/workspaces ?

Reply:
No.  This is just indicating what features each of the proposed backends could support.